x/authz/keeper/keeper.go

```diff
@@ -149,31 +149,33 @@ func (k Keeper) DispatchActions(ctx sdk.Context, grantee sdk.AccAddress, msgs []
 // SaveGrant method grants the provided authorization to the grantee on the granter's account
 // with the provided expiration time and insert authorization key into the grants queue. If there is an existing authorization grant for the
 // same `sdk.Msg` type, this grant overwrites that.
-func (k Keeper) SaveGrant(ctx sdk.Context, grantee, granter sdk.AccAddress, authorization authz.Authorization, expiration time.Time) error {
+func (k Keeper) SaveGrant(ctx sdk.Context, grantee, granter sdk.AccAddress, authorization authz.Authorization, expiration *time.Time) error {
 	store := ctx.KVStore(k.storeKey)
-	skey := grantStoreKey(grantee, granter, authorization.MsgTypeURL())
-	oldGrant, found := k.getGrant(ctx, skey)
+	msgType := authorization.MsgTypeURL()
+	skey := grantStoreKey(grantee, granter, msgType)
 
 	grant, err := authz.NewGrant(ctx.BlockTime(), authorization, expiration)
 	if err != nil {
 		return err
 	}
 
-	bz := k.cdc.MustMarshal(&grant)
-	store.Set(skey, bz)
-
-	if found {
-		// if expiration is not the same, remove old key and add the new key to queue
-		if !oldGrant.Expiration.Equal(expiration) {
-			if err := k.removeFromGrantQueue(ctx, skey, oldGrant.Expiration, granter, grantee); err != nil {
-				return err
-			}
-
-			k.insertIntoGrantQueue(ctx, granter, grantee, authorization.MsgTypeURL(), expiration)
+	var oldExp *time.Time
+	if oldGrant, found := k.getGrant(ctx, skey); found {
+		oldExp = oldGrant.Expiration
+	}
+	if oldExp != nil && (expiration == nil || !oldExp.Equal(*expiration)) {
+		if err = k.removeFromGrantQueue(ctx, skey, granter, grantee, *oldExp); err != nil {
+			return err
 		}
-	} else {
-		k.insertIntoGrantQueue(ctx, granter, grantee, authorization.MsgTypeURL(), expiration)
 	}
+	if expiration != nil && (oldExp == nil || !oldExp.Equal(*expiration)) {
```

Comment:
```suggestion
	if expiration != nil {
```

why not simply like this?

Reply:
If the expiration didn't change, then we didn't remove and we should not insert again. I will add a comment to the code.